Update composer check to be more granular

ryanrath opened the pull request:
Currently if any portion of `composer.json` is modified then `composer.lock`
must be modified or the process fails. Unfortunately, there are some times
when we would legitimately expect there to be no change in `composer.lock` when
`composer.json` is updated ( say, for instance, new entries are added to the
`classmap` section ).

It turns out that there are only a few areas of `composer.json` that we should
be inspecting to determine whether or not `composer.lock` should also be
updated.

This commit adds a new script `composer_check.py` that, when run, will compare
these sections for the current and original `composer.json`. If a difference is
found then its exit code will not be 0, else it will be 0.

`build.sh` has been updated to run this script and take the appropriate action
if there are changes detected.

ryanrath committed:
Update composer check to be more granular

Currently if any portion of `composer.json` is modified then `composer.lock`
must be modified or the process fails. Unfortunately, there are some times
when we would legitimately expect there to be no change in `composer.lock` when
`composer.json` is updated ( say, for instance, new entries are added to the
`classmap` section ).

It turns out that there are only a few areas of `composer.json` that we should
be inspecting to determine whether or not `composer.lock` should also be
updated.

This commit adds a new script `composer_check.py` that, when run, will compare
these sections for the current and original `composer.json`. If a difference is
found then its exit code will not be 0, else it will be 0.

`build.sh` has been updated to run this script and take the appropriate action
if there are changes detected.

jpwhite4 commented on travis/composer_check.py:
```diff
@@ -0,0 +1,85 @@
+    :return:
+    """
+    with open(path, 'r') as fh:
+        return json.loads(fh.read())
```
```suggestion
        return json.load(fh)
```

jpwhite4 committed:
Update travis/composer_check.py

Co-Authored-By: ryanrath <ryanrath@buffalo.edu>

jpwhite4 commented on travis/composer_check.py:
```diff
@@ -0,0 +1,85 @@
+
+# the scripts current working directory. Used in defining default values for
+# this script
+cwd = os.getcwd()
```
The python file `open()` will default to the current directory if you just specify a filename so there is no need to explicitly pull in the cwd and prepend it to the filename.

jpwhite4 commented on travis/composer_check.py:
```diff
@@ -0,0 +1,85 @@
+    new_content = read_json(new_file)
+
+    for key in ['require', 'require-dev', 'extra']:
+        left = value_or_else(orig_content, key)
```
This code will crash if one of the keys is missing from the content objects. You could mitigate this by having the value_or_else return an empty dict if the key is missing.

ryanrath committed:
Updating the value_or_else default value

Make sure that, if there is no explicitly defined default value then we return
an empty dict.

ryanrath committed:
Updates per @jpwhite4 code review

per @jpwhite4 `open` will default to using the `cwd` so explicitly retrieving it
is not needed.

jpwhite4 commented on travis/composer_check.py:
```diff
@@ -0,0 +1,82 @@
+        return json.load(fh)
+
+
+def value_or_else(data, key, default=None):
```
This capability already exists as a python builtin:

https://docs.python.org/2.7/library/stdtypes.html?highlight=get#dict.get

ryanrath replied:
*nod* that's what I get for not googling before writing.

jpwhite4 commented on travis/composer_check.py:
```diff
@@ -0,0 +1,82 @@
+        left = value_or_else(orig_content, key)
+        right = value_or_else(new_content, key)
+
+        (added, removed, modified) = dict_compare(left, right)
```
Do you need a dict_compare() function? how about:
```python
if left != right:
    exit(1)
```

ryanrath replied:
yup, that works

ryanrath committed:
Simplifications per @jpwhite4
